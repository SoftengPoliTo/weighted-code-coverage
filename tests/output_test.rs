use std::{
    env::temp_dir,
    path::{Path, PathBuf},
};

use insta::sorted_redaction;
use weighted_code_coverage::{GrcovFormat, Mode, WccRunner};

const PROJECT_PATH: &str = "./tests/seahorse/";
const SNAPSHOTS_PATH: &str = "./snapshots/output/";
const COVERALLS_PATH: &str = "./tests/seahorse/coveralls.json";
const COVDIR_PATH: &str = "./tests/seahorse/covdir.json";
const JSON_OUTPUT: &str = "wcc.json";

#[test]
fn test_output_coveralls_files() {
    compare(
        GrcovFormat::Coveralls(PathBuf::from(COVERALLS_PATH)),
        Mode::Files,
        "output_coveralls_files",
    );
}

#[test]
fn test_output_covdir_files() {
    compare(
        GrcovFormat::Covdir(PathBuf::from(COVDIR_PATH)),
        Mode::Files,
        "output_covdir_files",
    );
}

#[test]
fn test_output_coveralls_functions() {
    compare(
        GrcovFormat::Coveralls(PathBuf::from(COVERALLS_PATH)),
        Mode::Functions,
        "output_coveralls_functions",
    );
}

#[test]
fn test_output_covdir_functions() {
    compare(
        GrcovFormat::Covdir(PathBuf::from(COVDIR_PATH)),
        Mode::Functions,
        "output_covdir_functions",
    );
}

fn compare<A: Into<PathBuf> + Default>(
    json_format: GrcovFormat<A>,
    mode: Mode,
    snapshot_name: &str,
) {
    let output_dir = temp_dir();

    let output = WccRunner::<PathBuf>::new()
        .grcov_format(json_format)
        .mode(mode)
        .json_path(&output_dir.join(JSON_OUTPUT))
        .html_path(Some(output_dir))
        .run(Path::new(PROJECT_PATH))
        .unwrap();

    insta::with_settings!({
        snapshot_path => Path::new(SNAPSHOTS_PATH),
        prepend_module_to_snapshot => false,
    },{
        insta::assert_yaml_snapshot!(snapshot_name, output, { ".files" => sorted_redaction(), ".ignored_files" => sorted_redaction(), ".files.*.functions" => sorted_redaction() });
    });
}
