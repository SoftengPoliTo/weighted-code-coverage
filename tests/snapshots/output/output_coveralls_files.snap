---
source: tests/output_test.rs
expression: output
---
files:
  - name: examples/multiple_app.rs
    metrics:
      cyclomatic:
        wcc: 0
        crap: 63.8
        skunk: 20
        complexity: 7.5
        isComplex: true
      cognitive:
        wcc: 0
        crap: 9.8
        skunk: 7.1
        complexity: 2.7
        isComplex: true
      coverage: 0
  - name: examples/single_app.rs
    metrics:
      cyclomatic:
        wcc: 0
        crap: 7.8
        skunk: 6.2
        complexity: 2.3
        isComplex: true
      cognitive:
        wcc: 0
        crap: 3.1
        skunk: 3.6
        complexity: 1.3
        isComplex: true
      coverage: 0
  - name: src/app.rs
    metrics:
      cyclomatic:
        wcc: 76
        crap: 10.6
        skunk: 12.6
        complexity: 10.3
        isComplex: false
      cognitive:
        wcc: 67.9
        crap: 9.3
        skunk: 11.2
        complexity: 9.1
        isComplex: false
      coverage: 86.6
  - name: src/command.rs
    metrics:
      cyclomatic:
        wcc: 69.1
        crap: 8.1
        skunk: 9.9
        complexity: 7.9
        isComplex: false
      cognitive:
        wcc: 45.5
        crap: 6.7
        skunk: 8.2
        complexity: 6.5
        isComplex: true
      coverage: 84.6
  - name: src/context.rs
    metrics:
      cyclomatic:
        wcc: 78.3
        crap: 9.5
        skunk: 10.7
        complexity: 9.5
        isComplex: false
      cognitive:
        wcc: 60.4
        crap: 7.5
        skunk: 8.4
        complexity: 7.5
        isComplex: false
      coverage: 92.5
  - name: src/error.rs
    metrics:
      cyclomatic:
        wcc: 0
        crap: 76.5
        skunk: 23.1
        complexity: 9
        isComplex: true
      cognitive:
        wcc: 5.9
        crap: 2.4
        skunk: 3.1
        complexity: 1.2
        isComplex: true
      coverage: 5.9
  - name: src/flag.rs
    metrics:
      cyclomatic:
        wcc: 69.5
        crap: 7.8
        skunk: 8.9
        complexity: 7.8
        isComplex: false
      cognitive:
        wcc: 69.5
        crap: 4.3
        skunk: 4.9
        complexity: 4.3
        isComplex: false
      coverage: 91.6
  - name: src/help.rs
    metrics:
      cyclomatic:
        wcc: 100
        crap: 1.5
        skunk: 1.5
        complexity: 1.5
        isComplex: false
      cognitive:
        wcc: 100
        crap: 0
        skunk: 0
        complexity: 0
        isComplex: false
      coverage: 100
  - name: src/lib.rs
    metrics:
      cyclomatic:
        wcc: 100
        crap: 1
        skunk: 1
        complexity: 1
        isComplex: false
      cognitive:
        wcc: 100
        crap: 0
        skunk: 0
        complexity: 0
        isComplex: false
      coverage: 100
project:
  total:
    cyclomatic:
      wcc: 64.8
      crap: 9.4
      skunk: 11.8
      complexity: 8.517241379310345
      isComplex: false
    cognitive:
      wcc: 56
      crap: 6.6
      skunk: 8.5
      complexity: 6.149425287356322
      isComplex: true
    coverage: 77.2
  min:
    cyclomatic:
      wcc: 0
      crap: 1
      skunk: 1
      complexity: 1
      isComplex: true
    cognitive:
      wcc: 0
      crap: 0
      skunk: 0
      complexity: 0
      isComplex: true
    coverage: 0
  max:
    cyclomatic:
      wcc: 100
      crap: 76.5
      skunk: 23.1
      complexity: 10.3
      isComplex: true
    cognitive:
      wcc: 100
      crap: 9.8
      skunk: 11.2
      complexity: 9.1
      isComplex: false
    coverage: 100
  average:
    cyclomatic:
      wcc: 54.8
      crap: 20.7
      skunk: 10.4
      complexity: 6.3
      isComplex: true
    cognitive:
      wcc: 49.9
      crap: 4.8
      skunk: 5.2
      complexity: 3.6
      isComplex: true
    coverage: 62.4
ignored_files:
  - src/action.rs
